stan/math/rev/fun/lb_constrain.hpp

```diff
@@ -0,0 +1,379 @@
+  if (unlikely(lb_val == NEGATIVE_INFTY)) {
+    return ret_type(identity_constrain(x, lb));
+  } else {
+    if (!is_constant<T>::value && !is_constant<L>::value) {
+      arena_t<promote_scalar_t<var, T>> arena_x = x;
+      auto exp_x = to_arena(arena_x.val().array().exp());
+      arena_t<ret_type> ret = exp_x + lb_val;
+      lp += arena_x.val().sum();
+      reverse_pass_callback(
+          [arena_x, ret, lp, arena_lb = var(lb), exp_x]() mutable {
+            arena_x.adj().array() += ret.adj().array() * exp_x + lp.adj();
+            const double ret_adj_sum = ret.adj().sum();
+            arena_lb.adj() += ret_adj_sum;
+            lp.adj() += ret_adj_sum;
+          });
+      return ret_type(ret);
+    } else if (!is_constant<T>::value) {
+      arena_t<promote_scalar_t<var, T>> arena_x = x;
+      auto exp_x = to_arena(arena_x.val().array().exp());
+      arena_t<ret_type> ret = exp_x + lb_val;
+      lp += arena_x.val().sum();
+      reverse_pass_callback([arena_x, ret, exp_x, lp]() mutable {
+        arena_x.adj().array() += ret.adj().array() * exp_x + lp.adj();
+        lp.adj() += ret.adj().sum();
+      });
+      return ret_type(ret);
+    } else {
+      auto x_ref = to_ref(value_of(x));
+      arena_t<ret_type> ret = x_ref.array().exp() + lb_val;
+      lp += x_ref.sum();
+      reverse_pass_callback([ret, lp, arena_lb = var(lb)]() mutable {
+        const double ret_adj = ret.adj().sum();
+        arena_lb.adj() += ret_adj;
+      });
+      return ret_type(ret);
+    }
+  }
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a matrix of lower bounds
+ * elementwise to each input element.
+ *
+ * @tparam T A type inheriting from `EigenBase` or a `var_value` with inner type
+ * inheriting from `EigenBase`.
+ * @tparam L A type inheriting from `EigenBase` or a `var_value` with inner type
+ * inheriting from `EigenBase`.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_all_matrix_t<T, L>* = nullptr,
+          require_any_st_var<T, L>* = nullptr>
+inline auto lb_constrain(const T& x, const L& lb) {
+  using std::exp;
+  using ret_type = return_var_matrix_t<T, T, L>;
+  if (!is_constant<T>::value && !is_constant<L>::value) {
+    arena_t<promote_scalar_t<var, T>> arena_x = x;
+    arena_t<promote_scalar_t<var, L>> arena_lb = lb;
+    auto is_inf_lb = to_arena(
+        (arena_lb.val().array() != NEGATIVE_INFTY).template cast<double>());
+    auto precomp_x_exp = to_arena((arena_x.val().array() * is_inf_lb).exp());
+    arena_t<ret_type> ret = (is_inf_lb).select(
+        precomp_x_exp + arena_lb.val().array(), arena_x.val().array());
+    reverse_pass_callback(
+        [arena_x, arena_lb, ret, is_inf_lb, precomp_x_exp]() mutable {
+          arena_x.adj().array() += ret.adj().array() * precomp_x_exp;
+          arena_lb.adj().array() += ret.adj().array() * is_inf_lb;
+        });
```

Comment:
This is sort of a dirty trick, but since `exp(0) = 1` We can cast all of these checks to double and then in the forward pass multiply `x` by the check array. When we make `ret` we iterate over `is_inf_lb` and when it's true it uses the values we want and when false just returns `arena_x` at that value. 

But the nice thing is that we can then have no checks in the reverse pass and get nice SIMD instructions since we are just doing an elementwise multiplication

```cpp
arena_x.adj().array() += ret.adj().array() * precomp_x_exp;
```

Reply:
PRs introducing such dirty tricks should be accompanied by benchmarks proving they are actually faster than the alternative implementation.

Reply:
This doesn't seem like a dirty trick. Maybe just a trick lol. I'm happy with it.

Reply:
Actually this ended being best case the same speed and worst case a bit slower so I reverted it to just do the checks